dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/relationship/hibernate/HibernateRelationshipStore.java

```diff
@@ -134,25 +135,48 @@ public Relationship getByRelationship( Relationship relationship )
     }
 
     @Override
-    public Relationship getByRelationshipKey( String relationshipKeyAsString )
+    public List<String> getUidsByRelationshipKeys( List<String> relationshipKeyList )
     {
-        RelationshipKey relationshipKey = RelationshipKey.fromString( relationshipKeyAsString );
+        List<Object> c = getSession()
+            .createNativeQuery( new StringBuilder()
+                .append( "WITH relationshipitemuids AS ( " )
+                .append( "SELECT RI.relationshipitemid, coalesce(TE.uid, PI.uid, PSI.uid) as uid " )
+                .append( "FROM relationshipitem RI " )
+                .append(
+                    "LEFT JOIN trackedentityinstance TE ON TE.trackedentityinstanceid = RI.trackedentityinstanceid " )
+                .append( "LEFT JOIN programinstance PI ON PI.programinstanceid = RI.programinstanceid " )
+                .append(
+                    "LEFT JOIN programstageinstance PSI ON PSI.programstageinstanceid = RI.programstageinstanceid " )
+                .append( ") " )
+                .append( "SELECT R.uid " )
+                .append( "FROM relationship R " )
+                .append( "INNER JOIN relationshiptype RT ON RT.relationshiptypeid = R.relationshiptypeid " )
+                .append( "INNER JOIN relationshipitemuids F ON F.relationshipitemid = R.from_relationshipitemid " )
+                .append( "INNER JOIN relationshipitemuids T ON T.relationshipitemid = R.to_relationshipitemid " )
+                .append( "WHERE concat(RT.uid, '-', F.uid, '-', T.uid) IN (:keys) " )
+                .append( "OR (RT.bidirectional AND concat(RT.uid, '-', T.uid, '-', F.uid) IN (:keys)) " )
+                .toString() )
+            .setParameter( "keys", relationshipKeyList )
+            .getResultList();
+
+        return c.stream().map( String::valueOf ).collect( Collectors.toList() );
```

Comment:
i think this stream iteration can be avoided by using 
`<R> NativeQuery<R> createNativeQuery(String sqlString, Class<R> resultClass);`